# Численное интегрироование

# Постановка задач

1. Построить квадратурную формулу максимально возможной степени точности вида $$\int\limits_0^1 \dfrac{f(x)}{\sqrt{x}} dx \approx A_0 f(x_0) + A_1 f(x_1).$$

2. Определить алгебраическую степень точности указанной квадратурной формулы $$\int\limits_0^{1} f(x)dx\approx \dfrac{5}{18} f(\dfrac{1}{2}-\sqrt{\dfrac{3}{5}})+\dfrac{4}{9}f(\dfrac{1}{2})+\dfrac{5}{18} f(\dfrac{1}{2}+\sqrt{\dfrac{3}{5}}).$$

3. Используя правило Рунге, провести сравнительный анализ квадратурных формул левых прямоугольников и Симпсона на примере вычисления интеграла $$I = \int\limits_1^2 \dfrac{e^x}{x+1} dx.$$

4. Вычислить (используя квадратурную формула типа Гаусса) с точностью $\varepsilon = 10^{-4}$ интеграл $$I = \int\limits_0^5 e^{-x^2+x} dx.$$

5. Найти с точностью $\varepsilon= 10^{-4}$ решение уравнения $$\dfrac{1}{\sqrt{2\pi}}\int\limits_0^x e^{-t^2} dt = 0.3.$$

# Задача 1

Для построения квадратурной формулы понадобятся соотношения \begin{eqnarray}
\begin{cases}
\int\limits_a^b \rho(x) x^idx = \sum\limits_{k=0}^{n}A_kx^i_k,\quad i=\overline{0,m},\\
\int\limits_a^b \rho(x) x^{m+1}dx \ne \sum\limits_{k=0}^{n}A_kx^i_k;
\end{cases}
\end{eqnarray}
причем в этом случае говорят, что квадратурная формула **имеет алгебраическую степень точности равную m**.
	
Соответственно для построения квадратурной формулы необходимо построить систему из соотношений (1), по которой мы найдем неизвестные. Затем найти следующее выражение, при котором равенство нарушается.

Из условия следует, что неизвестными значениями являются $A_0, A_1$, $x_0, x_1$. Итого 4 неизвестных. Для их отыскания следует построить систему из 4-ех уравнений по формуле (1). Тогда в соответствии со степенями $x^i$ запишем систему
$$\begin{cases}
	x^0 : \int\limits_{0}^1 \dfrac{1}{\sqrt{x}} dx = A_0 + A_1,\\
	x^1 : \int\limits_{0}^1 \dfrac{x}{\sqrt{x}} dx = A_0x_0 + A_1x_1,\\
	x^2 : \int\limits_{0}^1 \dfrac{x^2}{\sqrt{x}} dx = A_0x_0^2 + A_1x_1^2,\\
	x^3 : \int\limits_{0}^1 \dfrac{x^3}{\sqrt{x}} dx = A_0x_0^3 + A_1x_1^3.\\
	\end{cases}$$ 
Вычислим значения интегралов из левой части уравнений:
$$\int\limits_{0}^1 \dfrac{1}{\sqrt{x}} dx = 2\sqrt{x}\Big|_0^1 = 2.$$
$$\int\limits_{0}^1 \dfrac{x}{\sqrt{x}} dx = \int\limits_0^1 \sqrt{x} = \dfrac{2}{3} x^{\dfrac{3}{2}}\Big|_0^1 = \dfrac{2}{3}.$$
$$\int\limits_{0}^1 \dfrac{x^2}{\sqrt{x}} dx =\int\limits_{0}^1 x\sqrt{x} dx = \dfrac{2}{5}x^{\dfrac{5}{2}}\Big|_ 0^1 = \dfrac{2}{5}.$$
$$\int\limits_{0}^1 \dfrac{x^3}{\sqrt{x}} dx =\int\limits_{0}^1 x^2\sqrt{x} dx = \dfrac{2}{7}x^{\dfrac{7}{2}}\Big|_ 0^1 = \dfrac{2}{7}.$$

В итоге имеем систему нелинейных уравнений $$\begin{cases} A_0 + A_1 = 2, \\ A_0x_0 + A_1x_1 = \dfrac23,\\ A_0x_0^2 + A_1x_1^2 = \dfrac{2}{5}, \\ A_0x_0^3 + A_1x_1^3 = \dfrac{2}{7}. \end{cases}$$
Первое, второе и третье уравнения домножим на $-x_0$ и прибавим ко второму, третьему и четвертому соответственно. Тогда будем иметь систему с тремя неизвестными 
$$
\begin{cases}
A_1x_1 - A_1x_0 = \dfrac23 - 2x_0,\\
A_1x_1^2 - A_1x_1x_0 = \dfrac{2}{5} - \dfrac23x_0,\\
A_1x_1^3 - A_1x_1^2x_0 = \dfrac{2}{7} - \dfrac{2}{5}x_0.
\end{cases}
$$
Поступаем аналогично, но умножим первое и второе уравнения на $-x_1$ и прибавим к второму и третьему соответственно. Тогда имеем более простую нелинейную систему из 2-ух уравнений 
$$
\begin{cases}
x_0 + x_1 - 3 x_0x_1 = \dfrac35,\\
x_0 + x_1 - \dfrac53 x_0x_1 = \dfrac57.
\end{cases}
$$
Вычтем из второго уравнения первое и получим $$\dfrac{4x_0x_1}{3} = \dfrac{4}{35}\Rightarrow x_0x_1 = \dfrac{3}{35}.$$
В предположении, что $x_1\ne 0$ (подстановкой в исходную систему можно проверить, что это действительно так), выражаем $x_0$:
$$x_0 = \dfrac{3}{35x_1}.$$
Подставляя это выражение в первое уравнение, получим $$\dfrac{3}{35x_1} + x_1 - 3\cdot \dfrac{3}{35} = \dfrac35.$$
Отсюда получаем квадратное уравнение $$x_1^2 - \dfrac{6}{7} x_1 + \dfrac{3}{35} = 0.$$
Корнями которого являются значения $$x_1 = \dfrac37 + \dfrac{2\sqrt{30}}{35},\quad x_1 = \dfrac37 - \dfrac{2\sqrt{30}}{35}.$$
Подстановкой каждого из значений в выражение $x_0 = \dfrac{3}{35x_1}$ можно получить, что $$x_0 = \dfrac37 - \dfrac{2\sqrt{30}}{35},\quad x_0 = \dfrac37 +\dfrac{2\sqrt{30}}{35}.$$
Таким образом, мы можем выбрать любое из двух значений. Обусловимся тем, что узлы мы строим в порядке возрастания. Поэтому в итоге выберем $$x_0 = \dfrac37 - \dfrac{2\sqrt{30}}{35},\quad x_1 = \dfrac37 +\dfrac{2\sqrt{30}}{35}.$$
Теперь, используя полученные значения, мы можем из первых двух уравнений исходной системы вычислить $A_0$ и $A_1$. Выпишем эти уравнения, подставив $x_0$, $x_1$:
$$\begin{cases}
A_0 + A_1 = 2,\\
\dfrac{(15-2\sqrt{30})A_0 + (15+2\sqrt{30})A_1}{35}=\dfrac23.
\end{cases}$$
Выразим из первого уравнения $A_1 = 2 - A_0$ и подставим во второе:
$$\dfrac{15A_0-2\sqrt{30}A_0 + 2(15+2\sqrt{30}) - 15A_0 - 2\sqrt{30}A_0}{35}=\dfrac23.$$
Преобразуем уравнение 
$$\dfrac{-4\sqrt{30}A_0}{35} = \dfrac23 - \dfrac{30}{35} - \dfrac{4\sqrt{30}}{35}.$$
$$A_0 = \dfrac{4\cdot 35}{21\cdot4\sqrt{30}}+1 = 1+\dfrac{\sqrt{30}}{18}.$$
Тогда $$A_1 = 2 - 1 - \dfrac{\sqrt{30}}{18}=1-\dfrac{\sqrt{30}}{18}.$$
Итого имеем $$A_0 = 1+\dfrac{\sqrt{30}}{18},\quad A_1 = 1-\dfrac{\sqrt{30}}{18},\quad x_0 = \dfrac37 - \dfrac{2\sqrt{30}}{35},\quad x_0 = \dfrac37 +\dfrac{2\sqrt{30}}{35}.$$
В итоге получаем квадратурную формулу, имеющую алгебраическую степень точности не ниже 3. Для уточнения построим еще одно уравнение и проверим, удовлетворяют ли полученные нами данные этому уравнению:
$$x^4 : \int\limits_{0}^1 \dfrac{x^4}{\sqrt{x}} dx = A_0x_0^4 + A_1x_1^4.$$
Как и прежде, вычисляем интеграл:
$$\int\limits_{0}^1 \dfrac{x^4}{\sqrt{x}} dx = \int\limits_{0}^1 x^3\sqrt{x} dx = \dfrac{2}{9}x^{\dfrac{9}{2}}\Big|_0^1) = \dfrac{2}{9}.$$
Тогда $$A_0x_0^4 + A_1x_1^4 = \dfrac29.$$
С помощью компьютерных вычислений посчитаем значение слева.

In [1]:
import math
A_0 = (18 + math.sqrt(30)) / 18
A_1 = (18 - math.sqrt(30)) / 18
x_0 = (15 - 2*math.sqrt(30)) / 35
x_1 = (15 + 2*math.sqrt(30)) / 35

A_0 * x_0**4 + A_1 * x_1**4

0.2106122448979593

In [2]:
A_0 * x_0**4 + A_1 * x_1**4 == 2/9

False

Таким образом, при степени 4 точность нарушается. Следовательно, мы построили квадратурную формулу с АСТ = 3 вида $$\int\limits_0^1 \dfrac{f(x)}{\sqrt{x}}dx = \dfrac{18 + \sqrt{30}}{18}\cdot f\left(\dfrac{15 - 2\sqrt{30}}{35}\right) + \dfrac{18 - \sqrt{30}}{18}\cdot f\left(\dfrac{15 + 2\sqrt{30}}{35}\right).$$

# Задача 2

Для решения задачи понадбятся соотношения (1). Таким образом необходимо строить по одному уравнению из соотношений (1) до тех пор, пока мы не получим неравенство.

Теперь обратим внимание на составляющие элементы формулы (1), а именно $\rho(x), A_k, x_k$. Квадратурная формула для нашей задачи записывается в общем случае в виде
$$I(f) = \int\limits_a^b \rho(x)f(x)dx \approx A_0 f(x_0) + A_1 f(x_1)+ A_2 f(x_2).$$
Из постановки задачи мы выделяем, что $$\rho(x) = 1,\quad A_0 =\dfrac5{18},\quad x_0 =\dfrac12-\sqrt{\dfrac35},\quad A_1 = \dfrac49,\quad x_1 = \dfrac12,\quad A_2=\dfrac5{18}, \quad x_2 =\dfrac12+\sqrt{\dfrac35}.$$
Остается лишь записывать соотношения из (1) последовательно, повышая степень $i$ на каждом шаге.

Берем $i=0$:
$$x^0 : \int\limits_0^1 1dx=1 \overset{?}{=}  A_0 + A_1+A_2.$$
Тогда $$1 \overset{?}{=} A_0 + A_1+A_2 = \dfrac5{18} + \dfrac49 + \dfrac5{18} = 1.$$
Равенство выполняется.

Берем $i=1$:
$$x^1 : \int\limits_0^{1} x dx = \dfrac12 \overset{?}{=}  A_0x_0 + A_1x_1+A_2x_2.$$
Тогда $$\dfrac12 \overset{?}{=} A_0x_0 + A_1x_1 +A_2x_2= \dfrac5{18}\cdot(\dfrac12-\sqrt{\dfrac35}) + \dfrac49\cdot(\dfrac12) + \dfrac5{18}\cdot(\dfrac12+\sqrt{\dfrac35}) = \dfrac12.$$
Равенство выполняется.

Берем $i=2$:
$$x^2 : \int\limits_0^{1} x^2 dx=\dfrac13 \overset{?}{=}  A_0x_0^2 + A_1x_1^2+A_2x_2^2,$$
Тогда $$\dfrac13 \overset{?}{=} A_0x_0^2 + A_1x_1^2+A_2x_2^2 = \dfrac5{18}\cdot(\dfrac12-\sqrt{\dfrac35})^2 + \dfrac49\cdot(\dfrac12)^2 + \dfrac5{18}\cdot(\dfrac12+\sqrt{\dfrac35})^2 = \dfrac7{12}.$$
Равенство не выполняется.

Таким образом, квадратурная формула имеет АСТ = 1.

# Задача 3

Импортируем все необходимые для дальнейшей работы библиотеки

In [3]:
import numpy as np
import pandas as pd
import math

Сперва сравним кваратурные формулы правых прямоугольников и трапеций без использования правила Рунге на примере вычисления интеграла $$I = \int\limits_1^2 \dfrac{e^x}{x+1} dx.$$

Обозначим подынтегральную функцию $$f(x) = \dfrac{e^x}{x+1}$$ и определим ее компьютерно

In [4]:
a, b = 1, 2
def f(x):
    return np.exp(x) / (x + 1)

## Сравнение простейших квадратурных формул

$\bullet$ *Формула $$I_\text{лп}(f) = (b-a)f(a) $$ называется **простейшей квадратурной формулой левых прямоугольников***.

Реализуем программно метод, вычисляющий это значение при заданных параметрах $f(x), a, b$.

In [5]:
def simple_right_rectangles(f, a, b):
    return (b - a) * f(a)

In [6]:
simple_right_rectangles(f, 1, 2)

1.3591409142295225

$\bullet$ *Формула $$R_\text{лп}(f) = \dfrac{(b-a)^2}{2}f'(\eta),\quad \eta\in [a,b]$$ определяет **остаток простейшей квадратурной формулы левых прямоугольников**.*

С помощью этой формулы оценим погрешность квадратурной формулы как 
$$|R(f)| \leq \left| \dfrac{(b-a)^2}{2}\right| \max_{x\in [a,b]} |f'(x)|,$$
причем $$f'(x) = \dfrac{xe^x}{(x+1)^2}.$$
Это значение можно грубо оценить как $$|f'(x)| \leq \left|\dfrac{xe^x}{(x+1)^2}\right| \underset{x=2}{\leq} 1.65.$$

Тогда $$|R(f)|\leq 0.825.$$

Иначе с помощью компьютерных методов можно получить максимальное значение производной:

In [7]:
def derivative_f(x):
    return (x*np.exp(x)) / (x+1)**2

M_1 = np.max(np.absolute(derivative_f(np.linspace(1, 2, 10000))))
M_1

1.6420124664290334

Тогда оценка для остатка:

In [8]:
abs((b - a) ** 2 / 2) * M_1

0.8210062332145167

Что является достаточно высокой погрешностью.

$\bullet$ *Формула $$I_\text{симп} (f)= \dfrac{b-a}{6}\left(f(a) + 4 f(\dfrac{a+b}2) + f(b)\right)$$ называется **простейшей квадратурной формулой Симпсона.***

Реализуем программно метод, вычисляющий это значение при заданных параметрах $f(x), a, b$.

In [9]:
def simple_simpson(f, a, b):
    return (b - a) / 6 * (f(a) + 4*f((a+b)/2) + f(b))

In [10]:
simple_simpson(f, 1, 2)

1.8321436877356627

$\bullet$ *Формула $$R_\text{тп}(f) = -\dfrac{(b-a)^5}{32}\dfrac{f^{(4)}(\eta)}{90}, \quad \eta\in [a,b]$$ определяет **остаток простейшей квадратурной формулы Симпсона.***

С помощью этой формулы оценим погрешность квадратурной формулы как 
$$|R(f)| \leq \left| \dfrac{(b-a)^5}{32\cdot90}\right| \max_{x\in [a,b]} |f^{(4)}(x)|,$$
причем $$f^{(4)}(x) = \dfrac{\left({x}^{4}+6\,{x}^{2}-8\,x+9\right)\,{e}^{x}}{{(x+1)^5}}.$$
C помощью компьютерных методов можно получить максимальное значение производной:

In [11]:
def fourth_derivative_f(x):
    return ((x**4+6*x**2-8*x+9)*np.exp(x))/(x+1)**5

M_2 = np.max(np.absolute(fourth_derivative_f(np.linspace(1, 2, 10000))))
M_2

1.0034520628177426

Тогда оценка погрешности:

In [12]:
abs((b - a) ** 5 / (32*90)) * M_2

0.0003484208551450495

Что является еще более низким значением, чем в предыдущем случае. 

Таким образом, более хорошая оценка (но все равно очень грубая) была получена с помощью метода Симпсона. Однако такой результат вычисления значения интеграла нас в любом случае не устраивает, поэтому перейдем к методу Рунге.

Но сперва, используя те формулы, которые у нас есть, попробуем априорно оценить то, насколько частей нам нужно будет разбить отрезок интегрирования при применении правила Рунге с помощью формул:

$$N_\text{лс}\geq \dfrac{(b-a)^2}{2\varepsilon}\cdot M_1,\quad M_1 = \underset{x\in[a,b]}{\max}|f'(x)|,$$
$$N_\text{симп} \geq \sqrt[4]{\dfrac{(b-a)^5}{2880\varepsilon}M_2},\quad M_2 = \underset{x\in[a,b]}{\max}|f^{(4)}(x)|.$$

Посмотрим, чему будут равны эти значения при $\varepsilon = 10^{-3}$:

In [13]:
epsilon = 1e-3
N_1 = (b - a) ** 2 / (2 * epsilon) * M_1
N_2 = math.sqrt(math.sqrt((b - a) ** 5 / (2880 * epsilon) * M_2))

In [14]:
N_1

821.0062332145167

In [15]:
N_2

0.7682915127159656

При формуле правых прямоугольников мы получили достаточно большое значение для числа разбиений, при формуле трапеций сильно меньше. Теперь смоделируем вычисление интеграла по правилу Рунге и проверим, такие ли получатся разбиения.

## Правило Рунге
Пусть имеет место разложение остатка составной квадратурной формулы $$R(h,f) = c h^m + o(h^{m+1}),\quad c = \operatorname{const}.$$
Обозначим точное значение интеграла через $I$, а через $I_h$ --- приближенное значение соответствующее составной квадратурной формуле с шагом $h$. По предположению $$I = I_h + ch^m + O(h^{m+1})\approx I_h + ch^m.$$
Применим составную квадратурную формулу при разных параметрах: возьмем два значения $h_1$ и $h_2$ и найдем $I_{h_1}, I_{h_2}$. Тогда $$\begin{cases}
I\approx I_{h_1} + c\cdot h_1^m,\\
I\approx I_{h_2} + c\cdot h_2^m.
\end{cases}$$
Отсюда легко найти главную часть погрешности равную $$R(h,f)\approx \dfrac{I_{h_2}-I_{h_1}}{h_1^m - h_2^m}h_1^m = \dfrac{I_{h_2}-I_{h_1}}{1 - \left(\frac{h_2}{h_1}\right)^m}.$$
Эта формула дает выражение для главной части остатка квадратурной формулы. Таким образом, мы можем указать алгоритм апостериорной оценки: мы подбираем шаги $h_1$ и $h_2$ таким образом, чтобы $$|R(h,f)|\leq \varepsilon,$$ тогда интеграл будет считаться вычисленным с заданной точностью.
     
Вычисленная величина главной части остатка позволяет уточнить приближенное значение интеграла $$I\approx I_{h_1} + \dfrac{I_{h_2}-I_{h_1}}{1 - \left(\frac{h_2}{h_1}\right)^m}.$$
Таким образом, мы повысим точность.

При проведении расчетов обычно вычисляются интегралы с выбранным шагом $h_1$ и шагом $h_2 = \dfrac{h_1}2$. Такой способ задания изменения параметров позволяет учесть уже сделанные вычисления и таким образом оптимизировать процесс вычисления интеграла с точки зрения арифметических операций.

Теперь весь описанный алгоритм реализуем компьютерно. Причем в начале выбирем $h_1 = b-a$, а $h_2 = \dfrac{h_1}{2}$. 

Построенная компьютерная функция приниамет на вход следующие параметры:

- $m$ -- алгебраическая степень точности метода;

- compound_quadrature_formula -- составная квадратурная формула, которую алгоритм будет использовать для вычислений интеграла;

- $f(x)$ -- подынтегральная функция;

- $a, b$ -- концы отрезка интегрирования;

- $\varepsilon$ -- точность, с которой должен быть вычислен интеграл;

- $N_{divisions} \in \{0,1\}$ -- переменная, которая отвечает за то, чтобы возвращать или не возвращать число делений отрезка проделанных алгоритмом. 

In [16]:
def runge_rule(m, compound_quadrature_formula, f, a, b, epsilon, N_divisions=False):
    h_1 = b-a
    h_2 = h_1 / 2
    iterations = 1
    I_h1 = compound_quadrature_formula(f, a, b, h_1)
    I_h2 = compound_quadrature_formula(f, a, b, h_2)
    R = (I_h2 - I_h1) / (1 - (h_2/h_1)**m)
    I = I_h1 + R
    while abs(R) >= epsilon:
        h_1 = h_2
        h_2 = h_1 / 2
        I_h1 = compound_quadrature_formula(f, a, b, h_1)
        I_h2 = compound_quadrature_formula(f, a, b, h_2)
        R = (I_h2 - I_h1) / (1 - (h_2/h_1)**m)
        I = I_h1 + R
        iterations += 1
    if N_divisions:
        return I, iterations
    return I

## Сравнение составных квадратурных формул

Теперь определим составные квадратурные формулы, которые мы будем использовать для вычисления.


$\bullet$ *Формула $$I_\text{лс}(f) = h\sum_{k=0}^{N-1}f(x_{k}) = h\sum_{k=0}^{N-1}f(a+kh)$$ называется **составной квадратурной формулой левых прямоугольников**.*

Она имеет алгебраическую степень точности равную 1.

Реализуем программно метод, вычисляющий эту сумму при заданных параметрах $f(x), a, b, h$.

In [17]:
def left_rectangles(f, a, b, h):
    I = 0
    N = int((b-a)/h)
    for k in range(N):
        I += f(a + k*h)
    I *= h
    return I

$\bullet$ *Формула $$I_\text{тс}(f) = \dfrac{h}{3}(f_0+f_N+2(f_2+f_4+\dots+f_{N-2})+4(f_1+f_3+\dots+f_{N-1})), \quad f_i=f(x_i)$$ называется **составной квадратурной формулой Симпсона**.*

Реализуем программно метод, вычисляющий эту сумму при заданных параметрах $f(x), a, b, h$.

In [18]:
def simps(f, a, b, h):
    I = f(a) + f(b)
    N = int((b-a)/h)
    x_k = a
    for k in range(2,N, 2):
        I += 2*f(a + k * h)
    for k in range(1,N, 2):
        I += 4*f(a + k * h)
    I *= h/3
    return I

Итак, задавая обеим формулам одни и те же значения точности $\varepsilon$, мы можем сравнивать для них скорость число делений отрезка $[a,b]$ и время работы алгоритма для каждой из формул.

Возьмем $\varepsilon = 10^{-3}$:

In [19]:
eps = 1e-3

In [20]:
%%time
runge_rule(1, left_rectangles, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 6 ms


(1.831891770210002, 11)

In [21]:
%%time
runge_rule(2, simps, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 0 ns


(1.8318294576260676, 2)

Оба алгоритма отработали достаточно быстро. Алгоритм Симпсона проделал меньше делений отрезка.

Причем заметим, что число разбиений отрезка между методами отличается на 9 единиц. К тому же значение алгоритма левых прямоугольников сильно меньше того, которое мы вычислили априорно.

Возьмем $\varepsilon = 10^{-4}$:

In [22]:
eps = 1e-4

In [23]:
%%time
runge_rule(1, left_rectangles, f, 1, 2, eps, True)

CPU times: total: 62.5 ms
Wall time: 53 ms


(1.8318918078564406, 14)

In [24]:
%%time
runge_rule(2, simps, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 0 ns


(1.8318877676375942, 3)

Алгоритм с методом левых прямоугольников отработал медленнее, чем с Симпсоном. Также алгоритм левых прямоугольников проделал больше делений отрезка.

Возьмем $\varepsilon = 10^{-5}$:

In [25]:
eps = 1e-5

In [26]:
%%time
runge_rule(1, left_rectangles, f, 1, 2, eps, True)

CPU times: total: 406 ms
Wall time: 418 ms


(1.8318918084446687, 17)

In [27]:
%%time
runge_rule(2, simps, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 0 ns


(1.8318915533643054, 4)

Алгоритм с методом левых прямоугольников снова показал себя медленнее. Деление отрезка интегрирования между методами отличается на 13 единиц.

Возьмем $\varepsilon = 10^{-6}$:

In [28]:
eps = 1e-6

In [29]:
%%time
runge_rule(1, left_rectangles, f, 1, 2, eps, True)

CPU times: total: 6.72 s
Wall time: 6.7 s


(1.831891808454127, 21)

In [30]:
%%time
runge_rule(2, simps, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 500 µs


(1.8318917924697014, 5)

Можно заметить, что только метод левых прямоугольников стал работать гораздо медленее (от 6 секунд). Также метод Симпсона оказался сильно быстрее метода левых прямоугольников. Разбиение отрезка отличается уже на 16.

Возьмем $\varepsilon = 10^{-7}$:

In [31]:
eps = 1e-7

In [32]:
%%time
runge_rule(1, left_rectangles, f, 1, 2, eps, True)

CPU times: total: 56.1 s
Wall time: 56.2 s


(1.8318918084532716, 24)

In [33]:
%%time
runge_rule(2, simps, f, 1, 2, eps, True)

CPU times: total: 0 ns
Wall time: 0 ns


(1.8318917924697014, 5)

Таким образом, по результатам тестирования можно считать, что алгоритм Рунге с использованием метода левых прямоугольников оказался хуже алгоритма, использующего метод Симпсона, по крайней мере по скорости работы. Причем заметно, что чем выше точность, тем более явно проявляется тенденция, что алгоритм с методом левых прямоугольников работает медленнее алгоритма с методом Симпсона.

# Задача 4

Имеем интеграл $$I = \int\limits_0^5 e^{-x^2+x} dx$$
и точность $\varepsilon = 10^{-4}$. Для вычисления значения интеграла воспользуемся квадратурными формулами наивысшей алгебраической степени точности, или квадратурные формулы типа Гаусса.

Но без дополнительных преобразований применить к вычислению данного интеграла квадратурные формулы типа Гаусса невозможно. Из теории квадратур следует, что для применения формул Гаусса или Эрмита в случае произвольного отрезка $[a,b]$ необходимо использовать линейное преобразование $$x = \dfrac{b-a}{2}t + \dfrac{a+b}{2},\quad t \in [-1,1].$$
Тогда $x$ будет изменяться на отрезке $[a,b]$. Остаток квадратурной формулы для промежутка $[a,b]$ выражается через остаток на отрезке $[-1,1]$ следующим образом: $$R_n(f) = \left(\dfrac{b-a}{2}\right)^{2n+3}R_n(f).$$

Таким образом, в нашем случае применима замена вида
$$x = 2.5t + 2.5,\quad t \in [-1,1].$$
Отсюда $$t = \dfrac{x}{2.5}-1=0.4x-1, \quad x \in [0,5].$$
Подставим эту замену в рассматриваемый интеграл и получим $$I = \dfrac{b-a}2\int\limits_{-1}^1e^{-6.75t^2-10t-3.75}dt.$$

$\bullet$ Формула $$I(f) = \dfrac{b-a}2\int\limits_{-1}^1f(x)dx \approx \dfrac{b-a}2\sum_{k=0}^{n}A_k f(x_k)$$
Системой многочленов ортогональных по весу $p(x)=1$ на отрезке [-1, 1] является система многочленов Лежандра:
$$P_n(x)=\dfrac1{2^nn!}\cdot\dfrac{d^n}{dx^n}(x^2-1)^n.$$

с узлами $x_k$, которые являются корнями $$P_{n+1}(x)=0$$ и коэффициентами $$A_k = \dfrac2{(1-x_k)^2\left[P'_{n+1}(x_k)\right]^2},\quad k=\overline{0,n}$$ и называется __квадратурной формулой Гаусса__.

Можно получить выражение для остатка этой формулы $$R_n(f) = \dfrac{2^{2n+3}}{(2n+3)(2n+2)!}\left[\dfrac{(n+1)!^2}{(2n+2)!}\right]^2\cdot f^{(2n+2)}(\eta),\quad \eta\in [-1,1].$$

Причем $f(x) = e^{-6.75x^2-10x-3.75}$. В итоге для конкретно нашего случая квадратурная формула выглядит следующим образом:
$$I(f) = \dfrac{b-a}2\int\limits_{-1}^1f(x)dx \approx  2.5\sum_{k=0}^{n}A_k e^{-6.75x_k^2-10x_k-3.75}.$$

C учетом сделанной замены остаток выражается как $$R_n(f) = \left(\dfrac{b-a}{2}\right)^{2n+3}R_n(f) = 2.5^{2n+3}R_n(f) =\dfrac{5^{2n+3}}{(2n+3)(2n+2)}\left[\dfrac{(n+1)!^2}{(2n+2)!}\right]^2\cdot f^{(2n+2)}(\eta),\quad \eta\in [-1,1].$$

Но нам нужно еще оценить это значение:
$$|R_n(f)| \leq \left|\dfrac{5^{2n+3}}{(2n+3)(2n+2)}\left[\dfrac{(n+1)!^2}{(2n+2)!}\right]^2\cdot f^{(2n+2)}(\eta)\right| \leq M\cdot \left|\dfrac{5^{2n+3}}{(2n+3)(2n+2)}\left[\dfrac{(n+1)!^2}{(2n+2)!}\right]^2\right|,\quad \eta\in [-1,1],$$
где $$M = \max_{\eta \in [-1,1]}|f^{(2n+2)}(\eta)| = \max_{\eta \in [-1,1]}|( e^{-6.75\eta^2-10\eta-3.75}\sqrt{1-\eta^2})^{(2n+2)}|.$$

Аналитически рассчитывать это значение будет сложно. Поэтому предлагается использовать в качестве аналога для значения погрешности значение $$\Delta = |I(f, n) - I(f, n-1)| < \varepsilon.$$ То есть погрешность мы будем оценивать как разность приближенных значений интеграла при $n$ и $n-1$ узлах.

Таким образом, нам остается лишь реализовать программно процесс вычисления интеграла по формуле (1), оценивая погрешность по формуле (2). Но нам не задано количество узлов $n$, поэтому, перебирая $n=1,2,\ldots$, мы будем оценивать прогрешность вычисления интеграла на каждой итерации. Перебор будем производить до тех пор, пока не достигнем заданной точности.

Реализуем компьютерный алгоритм, который принимает на вход функцию $f(x)$ и точность $\varepsilon$ и возвращает приближенное значение интеграла и число узлов, которое понадобилось для достижения заданной точности.

In [34]:
import cmath
epsilon = 10e-4

def f(x):
    return cmath.exp(-6.75*x**2-10*x-3.75)

In [35]:
import sympy as sp
import numpy as np
import math

def P(n):
    y = sp.Symbol('y')
    pol = (1/((2**n)*math.factorial(n)))*sp.diff((y**2-1)**n, y, n)
    return pol

def subs(x, pol):
    y = sp.Symbol('y')
    return pol.subs(y, x)

def find_roots(pol):
    y = sp.Symbol('y')
    pol=sp.simplify(pol)
    roots = sp.solve(pol, y)
    return roots

In [36]:
def create_matrix(x):
    matrix = np.zeros((len(x), len(x)+1), dtype=np.complex128)

    for i in range(len(x)):
        for j in range(len(x)):
            matrix[i, j] = x[j] ** i
        if i % 2 != 0:
            matrix[i, len(x)] = 0
        else:
            matrix[i, len(x)] = 2 / (i + 1)
    return np.linalg.solve(matrix[:, :-1], matrix[:, -1])

In [37]:
def gauss_quadrature_formula(f, epsilon):
    y = sp.Symbol('y')
    n = 1
    I_new = np.inf
    I_last = 0
    while abs(I_new - I_last) >= epsilon:
        I_last = I_new
        I_new = 0
        pol=P(n)
        def_pol=sp.diff(pol, y, 1)
        arr_x=find_roots(pol)
        roots = create_matrix(arr_x)
        for i in range(len(arr_x)):
            I_new+=roots[i]*f(arr_x[i])
        I_new*=2.5
        n += 1
    return I_new, n-1

In [38]:
gauss_quadrature_formula(f, epsilon)

(1.7302344334799634, 7)


Таким образом, мы вычислили приближенное значение интеграла, и оно равно $$I \approx 1.7302.$$ Для этого нам понадобилось выбрать на отрезке $n=7$ узлов.

# Задача 5

По условию нам дано уравнение $$\dfrac1{\sqrt{2\pi}}\int\limits_0^x e^{-t^2}dt = 0.3.$$
Введем обозначение $$f(x) = \dfrac1{\sqrt{2\pi}}\int\limits_0^x e^{-t^2}dt - 0.3$$
Тогда поставленная задача сводится к отысканию решения нелинейного уравнения $$f(x) = 0.$$
Решением подобного рода задач мы занимались ранее, поэтому все подробные выкладки из теории, касающиеся нелинейных уравнений, теперь будут опускаться.

## Отделение корня

Прежде чем находить приближенное значение корня, нам необходимо провести процедуру отделения корня. Вспомним, что на отрезке $[a,b]$ существует единственный корень уравнения $f(x) = 0$, если

- функция $f(x)$ на концах отрезка имеет разные по знаку значения;

- функция $f(x)$ на отрезке монотонна.

### Монотонность функции
Начнем со второго пункта, так как его можно быстро обосновать. Рассмотрим производную функции $$f'(x) = \dfrac1{\sqrt{2\pi}}e^{-x^2}$$ как производная интеграла с переменным верхним пределом (по теореме Барроу). Функция $f'(x)$ состоит из функции $e^{-x^2}$, которая, в свою очередь, является строго положительной на множестве $[0, +\infty)$. Тогда $$f'(x) > 0,\quad x \in [0, +\infty),$$
а значит исходная функция монотонна на этом множестве.

Перейдем к первому пункту. Чтобы определять знак значения функции $f(x)$ в точке, нам необходимо вычислять интеграл, который присутствует в функции. Поэтому далее мы отдельно рассмотрим алгоритм вычисления подобных интегралов.

### Приближенное вычисление интеграла $\int\limits_0^x e^{-t^2}dt$

Сперва определим метод для вычисления значения подынтегральной функции $g(t) = e^{-t^2}$:

In [39]:
def g(t):
    return np.e**(-t**2)

Для вычисления интеграла при заданной точности воспользуемся правилом Рунге для составных формул средних прямоугольников и Симпсона (поскольку они не были реализованы в предыдущих заданиях) и сравним по времени: какая из формул быстрее будет вычислять значение интеграла, такую и будем дальше применять.

$\bullet$ *Формула $$I_\text{сс}(f) = h\sum_{k=0}^{N-1}f\left(\dfrac{x_k+x_{k+1}}{2}\right) = h\sum_{k=0}^{N-1}f\left(a+\left(k+\dfrac12\right)h\right)$$ называется **составной квадратурной формулой средних прямоугольников**.

Реализуем программно метод, вычисляющий сумму при заданных параметрах $f(x), a, b, h$.

In [40]:
def avg_rectangles(f, a, b, h):
    I = 0
    N = int((b-a)/h)
    for k in range(N):
        I += f(a + (k+1/2)*h)
    I *= h
    return I

$\bullet$ *Формула $$I_\text{симп с}(f) = \dfrac{h}{3}\big(f_0 + f_N + 2(f_2 + f_4 + \ldots + f_{N-2}) + 4(f_1 + f_3 + \ldots + f_{N-1})\big),\quad f_i = f(x_i)$$ называется **составной квадратурной формулой Симпсона**.*

Реализуем программно метод, вычисляющий сумму при заданных параметрах $f(x), a, b, h$.

In [41]:
def simpson(f, a, b, h):
    x_k = a
    I = f(x_k)
    N = int((b-a)/h)
    for i in range(1, N):
        x_k += h
        if i % 2 == 1:
            I += 4 * f(x_k)
        else:
            I += 2 * f(x_k)
    I += f(x_k + h)
    I *= h/3
    return I

Сравним работу двух квадратурных формул на примере вычисления значения интеграла $$I = \int\limits_0^2 e^{-t^2}dt$$ с точностью $\varepsilon = 10^{-7}$.

In [42]:
%%time
runge_rule(1, avg_rectangles, g, 0, 2, 1e-7)

CPU times: total: 0 ns
Wall time: 2.01 ms


0.8820813674730624

In [43]:
%%time
runge_rule(3, simpson, g, 0, 2, 1e-7)

CPU times: total: 0 ns
Wall time: 0 ns


0.8820813951998405

Как можно заметить, алгоритм с методом Симпсона в качестве квадратурной формулы отработал гораздо быстрее алгоритма с формулой средних прямоугольников. Следовательно, для дальнейших вычислений отдадим предпочтение алгоритму с использованием метода Симпсона.

### Вычисление значений интеграла на концах отрезка

Напомним, что нас интересует отрезок, на котором функция $$f(x) = \dfrac1{\sqrt{2\pi}}\int\limits_0^x e^{-t^2}dt - 0.3$$ будет иметь на концах разные по знаку значения.

Определим компьютерный метод, который будет применять построенный ранее алгоритм Симпсона при заданной точности для вычисления значения функции $f(x)$ в заданной точке.

In [44]:
def f(x, epsilon):
    return ( 1 / np.sqrt(2 * np.pi))*runge_rule(3, simpson, g, 0, x, epsilon) - 0.3

Пусть значение интеграла в функции $f(x)$ будет вычисляться с точностью $\varepsilon=10^{-9}$, чтобы остался еще запас по погрешности для дальнейших вычислений.

In [45]:
epsilon = 1e-9

Возьмем точку $x = 2$ и рассмотрим значение в ней:

In [46]:
f(2, epsilon)

0.0518995615373411

То есть $$f(2) > 0.$$
Поскольку подынтегральная функция монотонно убывающая на $[0, +\infty)$, все значения которые справа точно дадут положительный результат (так как она строго положительна). $\dfrac1{\sqrt{2\pi}} \approx 0.398$, следовательно значение интеграла должно быть чуть меньше 1.

Возьмем значения из промежутка $(0,2)$. Например, $x = 1$.

In [47]:
f(1, epsilon)

-0.0020602774108717248

Таким образом, $$f(1) < 0.$$
В итоге мы определили отрезок $[1, 2]$, на котором функция имеет различные значения на концах и монотонна. Следовательно, на этом отрезке находится единственный корень исследуемого уравнения.

Также заметим, что из-за монотонности подынтегральной функции, функция $f(x)$ больше нигде кроме выбранного отрезка менять знак не будет. То есть то решение уравнения, которое мы найдем на данном отрезке, будет единственным на всем множестве.

На этом процедуру отделения корня мы не закончим. Мы уточним отрезок, применив к нему **метод дихотомии** (метод деления отрезка пополам). Сильно увлекаться не будем, и уточним лишь до значения $\varepsilon_2 = 10^{-1}$.

In [49]:
a, b = 1, 2
epsilon_2 = 1e-1

dichotomy_table = [[a, b, b - a, f(a, epsilon), f(b, epsilon) , (a + b)/2]]
c = 0
while b - a > epsilon_2:
    c = (a + b) / 2
    if f(c, epsilon) * f(a, epsilon) >= 0:
        a = c
    else:
        b = c
    dichotomy_table.append([a, b, b - a, f(a, epsilon), f(b, epsilon) , (a + b) / 2])
    
pd.DataFrame(dichotomy_table, columns = ['a', 'b', 'b-a', 'f(a)', 'f(b)', '(a+b)/2'])

a       b     b-a      f(a)      f(b)  (a+b)/2
0  1  2.0000  1.0000 -0.002066  0.054748  1.50000
1  1  1.5000  0.5000 -0.002066  0.042084  1.25000
2  1  1.2500  0.2500 -0.002066  0.026403  1.12500
3  1  1.1250  0.1250 -0.002066  0.014121  1.06250
4  1  1.0625  0.0625 -0.002066  0.006558  1.03125

Таким образом, применив метод дихотомии, мы получили уточненный отрезок $[1, 1.0625]$, на котором находится корень исследуемого уравнения.

Теперь мы можем применять методы решения нелинейных уравнений для уточнения приближенного значения корня.

## Метод Ньютона 

Формула этого метода имеет следующий вид $$x^{k+1} = x^k - \dfrac{f(x^k)}{f'(x^0)},\quad k=0,1,\ldots;\quad x^0.$$

**Условия сходимости метода:**

Необходимо выбрать начальное приближение $x_0$ так, чтобы выполнялись условия:

1. Функция $f(x)$ определена и дважды непрерывно дифференцируема на отрезке $$s_0 = [x^0; x^0 + 2h_0],\quad h_0 =- \dfrac{f(x^0)}{f'(x^0)}.$$ При этом на концах отрезка $f(x)f'(x)\ne 0$.

2. Для начального приближения $x^0$ выполняется неравенство $$2|h_0|M \leq |f'(x^0)|,\quad M = \underset{x\in s_0}{\max}|f''(x)|.$$

Итак, на данный момент в нашем случае мы имеем следующие значения:

- $f(x) = \dfrac1{\sqrt{2\pi}}\int\limits_0^x e^{-t^2}dt - 0.3;$

- $g(x) = e^{-x^2};$

- $x^* \in [1, 1.0625]$.

Выбираем $x^0 = 1$. Проверим выполнение первого условия:

In [50]:
x_0 = 1
print('x_0 =', x_0) 
h_0 = - f(x_0, epsilon) / (( 1 / np.sqrt(2 * np.pi))*g(x_0))
print('h_0 =', h_0) 
s_0 = np.linspace(x_0, x_0 + 2*h_0, 10000)
print('s_0 = [', s_0[0], ';', s_0[-1], ']')

x_0 = 1
h_0 = 0.014079444532643681
s_0 = [ 1.0 ; 1.0281588890652873 ]


Проверяем, чтобы на концах отрезка значение функции и ее производной не обращались в ноль одновременно:

In [51]:
f(s_0[0], epsilon)* ( 1 / np.sqrt(2 * np.pi))*g(s_0[0])

-0.000303261088203498

In [52]:
f(s_0[-1], epsilon)*( 1 / np.sqrt(2 * np.pi))*g(s_0[-1])

0.0002711076054324562

Значения в ноль не обращаются, условие выполнено. 

Проверим второе условие сходимости. Для него нам нужна вторая производная $f''(x)$ или же $\dfrac1{\sqrt{2\pi}}g'(x)$:
$$f''(x) = \dfrac1{\sqrt{2\pi}}g'(x) = -2\dfrac1{\sqrt{2\pi}}xe^{-x^2}.$$
Получившаяся функция является линейной комбинацией монотонно убывающих непрерывных на выбранном отрезке функций, а следовательно, сама является непрерывной и монотонно убывающей на этом отрезке. Тогда ее наибольшее значение достигается на левом конце рассматриваемого отрезка.

In [53]:
def derivative_g(x):
    return ( 1 / np.sqrt(2 * np.pi))*np.e ** (-x ** 2)*(-2)*x

Применяя компьютерные методы, найдем число $M$, которое требуется во втором условии сходимости:

In [54]:
M = np.max(np.absolute(derivative_g(s_0)))
M

0.29352532634747985

Также с помощью компьютерных методов проверяем выполнение самого условия:

In [55]:
2 * np.absolute(h_0) * M <= np.absolute(g(x_0))

True

Итого оба условия сходимости выполнены, следовательно, построенный нами по методу Ньютона итерационный процесс будет сходиться к нашему искомому корню.

Теперь компьютерно реализуем сам итерационный процесс. Введем заданную по условию точность $\epsilon_3 = 10^{-4}$:

In [56]:
epsilon_3 = 1e-4

Определим функцию $$\varphi(x) = x - \dfrac{f(x)}{f'(x)}$$

In [57]:
def phi(x):
    return x - f(x, epsilon) / (( 1 / np.sqrt(2 * np.pi))*g(x))

С помощью цикла реализуем итерационный процесс по формуле $$x^{k+1} = \varphi(x^k),\ \texttt{while}\ |x_{k+1} - x_k| > \varepsilon,\ k=0,1,\ldots$$
Выпишем в виде таблицы итоговый результат.

In [58]:
x_k = x_0
x_k1 = phi(x_0)
iterations = [[x_k1, np.absolute(x_k1 - x_k)]]
while np.absolute(x_k1 - x_k) >= epsilon_3:
    x_k = x_k1
    x_k1 = phi(x_k)
    iterations.append([x_k1, np.absolute(x_k1 - x_k)])

In [59]:
newton_table = pd.DataFrame(iterations, columns=[('Метод Ньютона', 'x_k'), ('Метод Ньютона', '|x_k+1 - x_k|')])
newton_table.columns = pd.MultiIndex.from_tuples(newton_table.columns, names=['Назание метода', ''])
newton_table

Назание метода Метод Ньютона              
                         x_k |x_k+1 - x_k|
0                   1.014079  1.407944e-02
1                   1.014267  1.875341e-04
2                   1.014267  1.858565e-07

Итак, мы достигли заданной точности за 3 итерации. Причем полученная в итоге погрешность оказалась гораздо меньше, чем нам нужна была. 

Таким образом, мы нашли корень уравнения $$f(x) = \dfrac1{\sqrt{2\pi}}\int\limits_0^x e^{-t^2}dt - 0.3 = 0$$ с точностью $10^{-4}$. Убедимся в том, что это действительно корень:

In [66]:
f(1.014267, epsilon)

2.9551841029817894e-08

То есть полученное значение действительно оказалось корнем уравнения точности $\varepsilon = 10^{-8}$.

Также можно заметить, что, подставив значение $x_{k+1}$ из последней итерации, мы получим

In [67]:
f(x_k1, epsilon)

-3.134420500927604e-11

То есть, несмотря на поставленное условие, за 3 итерации мы смогли достичь достаточно хорошей точности для корня уравнения. Получившийся корень имеет вид:

In [68]:
x_k1

1.0142667927967344

Корень исходного уравнения приблизительно равен $$x^* \approx 1.0142667927967344.$$